{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ae835d6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import logging\n",
    "import dataclasses\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import logging\n",
    "sys.path.append('..')\n",
    "\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c64cb5be",
   "metadata": {},
   "outputs": [],
   "source": [
    "from rnacloud_genome_reference.common.gtf import GTFHandler, SpliceJunctionPosition\n",
    "from rnacloud_genome_reference.common.gnomad import GenomicRegion, query_genomic_regions, GenomicRegionQuerier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c39b1b53",
   "metadata": {},
   "outputs": [],
   "source": [
    "logging.basicConfig(level=logging.DEBUG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f5468883",
   "metadata": {},
   "outputs": [],
   "source": [
    "pc = pd.read_csv(\"../temp/clinically_significant_protein_coding_genes.tsv\", sep=\"\\t\", low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "bc085772",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_handler = GTFHandler(\"../tests/fixtures/GCF_000001405.40_GRCh38.p14_genomic.sorted.gtf.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4299a4c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a1169c39",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "def convert_sj_positions_to_genomic_region(sj_position: SpliceJunctionPosition) -> GenomicRegion:\n",
    "    return GenomicRegion(\n",
    "        chrom='11',\n",
    "        start=sj_position.pos,\n",
    "        stop=sj_position.pos\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2ecb9029",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:rnacloud_genome_reference.common.gtf:Fetching transcript for Entrez Gene ID: 9636 in primary region NC_000001.11:1013497-1014540\n",
      "INFO:rnacloud_genome_reference.common.gtf:Obtaining transcript for Entrez Gene ID: 9636 at location NC_000001.11:1013497-1014540 (MANE: True)\n",
      "DEBUG:rnacloud_genome_reference.common.gtf:Obtained transcript: NM_005101.4\n",
      "INFO:rnacloud_genome_reference.common.gtf:Obtaining exons for transcript: NM_005101.4 at location NC_000001.11:1013497-1014540\n",
      "DEBUG:rnacloud_genome_reference.common.gtf:Obtained Exons: [Exon(chromosome='NC_000001.11', start=1013497, end=1013576, strand='+', sequence=None, exon_no=1), Exon(chromosome='NC_000001.11', start=1013984, end=1014540, strand='+', sequence=None, exon_no=2)]\n",
      "INFO:rnacloud_genome_reference.common.gnomad:Initialized GenomicRegionQuerier with endpoint: https://gnomad.broadinstitute.org/api\n",
      "INFO:rnacloud_genome_reference.common.gnomad:Querying 4 genomic regions\n",
      "DEBUG:rnacloud_genome_reference.common.gnomad:Adding region 1: 11:1013577-1013577\n",
      "DEBUG:rnacloud_genome_reference.common.gnomad:Adding region 2: 11:1013578-1013578\n",
      "DEBUG:rnacloud_genome_reference.common.gnomad:Adding region 3: 11:1013982-1013982\n",
      "DEBUG:rnacloud_genome_reference.common.gnomad:Adding region 4: 11:1013983-1013983\n",
      "DEBUG:rnacloud_genome_reference.common.gnomad:Generated GraphQL query. No. of query parts: 4\n",
      "INFO:gql.transport.requests:>>> {\"query\": \"query MultipleRegions {\\n  region1: region(\\n    chrom: \\\"11\\\"\\n    start: 1013577\\n    stop: 1013577\\n    reference_genome: GRCh38\\n  ) {\\n    variants(dataset: gnomad_r4) {\\n      chrom\\n      pos\\n      alt\\n      lof_filter\\n      joint {\\n        ac\\n        an\\n        hemizygote_count\\n        homozygote_count\\n        filters\\n      }\\n    }\\n  }\\n  region2: region(\\n    chrom: \\\"11\\\"\\n    start: 1013578\\n    stop: 1013578\\n    reference_genome: GRCh38\\n  ) {\\n    variants(dataset: gnomad_r4) {\\n      chrom\\n      pos\\n      alt\\n      lof_filter\\n      joint {\\n        ac\\n        an\\n        hemizygote_count\\n        homozygote_count\\n        filters\\n      }\\n    }\\n  }\\n  region3: region(\\n    chrom: \\\"11\\\"\\n    start: 1013982\\n    stop: 1013982\\n    reference_genome: GRCh38\\n  ) {\\n    variants(dataset: gnomad_r4) {\\n      chrom\\n      pos\\n      alt\\n      lof_filter\\n      joint {\\n        ac\\n        an\\n        hemizygote_count\\n        homozygote_count\\n        filters\\n      }\\n    }\\n  }\\n  region4: region(\\n    chrom: \\\"11\\\"\\n    start: 1013983\\n    stop: 1013983\\n    reference_genome: GRCh38\\n  ) {\\n    variants(dataset: gnomad_r4) {\\n      chrom\\n      pos\\n      alt\\n      lof_filter\\n      joint {\\n        ac\\n        an\\n        hemizygote_count\\n        homozygote_count\\n        filters\\n      }\\n    }\\n  }\\n}\"}\n",
      "DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): gnomad.broadinstitute.org:443\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[GenomicRegion(chrom='11', start=1013577, stop=1013577, id=None), GenomicRegion(chrom='11', start=1013578, stop=1013578, id=None), GenomicRegion(chrom='11', start=1013982, stop=1013982, id=None), GenomicRegion(chrom='11', start=1013983, stop=1013983, id=None)]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:urllib3.connectionpool:https://gnomad.broadinstitute.org:443 \"POST /api HTTP/1.1\" 400 83\n",
      "INFO:gql.transport.requests:<<< {\"errors\":[{\"message\":\"Query is too expensive (44). Maximum allowed cost is 25.\"}]}\n",
      "ERROR:rnacloud_genome_reference.common.gnomad:Failed to execute GraphQL query: {'message': 'Query is too expensive (44). Maximum allowed cost is 25.'}\n"
     ]
    },
    {
     "ename": "TransportQueryError",
     "evalue": "{'message': 'Query is too expensive (44). Maximum allowed cost is 25.'}",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mTransportQueryError\u001b[39m                       Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[20]\u001b[39m\u001b[32m, line 17\u001b[39m\n\u001b[32m     13\u001b[39m genomic_regions = [convert_sj_positions_to_genomic_region(sj) \u001b[38;5;28;01mfor\u001b[39;00m sj \u001b[38;5;129;01min\u001b[39;00m a]\n\u001b[32m     15\u001b[39m \u001b[38;5;28mprint\u001b[39m(genomic_regions)\n\u001b[32m---> \u001b[39m\u001b[32m17\u001b[39m gnomad_frequency = \u001b[43mquery_genomic_regions\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgenomic_regions\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     19\u001b[39m \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/workspaces/rnacloud_genome_reference/misc/../rnacloud_genome_reference/common/gnomad.py:158\u001b[39m, in \u001b[36mquery_genomic_regions\u001b[39m\u001b[34m(regions, endpoint_url, reference_genome, dataset, headers)\u001b[39m\n\u001b[32m    156\u001b[39m \u001b[38;5;66;03m# Create querier and execute query\u001b[39;00m\n\u001b[32m    157\u001b[39m querier = GenomicRegionQuerier(endpoint_url, reference_genome, dataset, headers)\n\u001b[32m--> \u001b[39m\u001b[32m158\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mquerier\u001b[49m\u001b[43m.\u001b[49m\u001b[43mquery_multiple_regions\u001b[49m\u001b[43m(\u001b[49m\u001b[43mregions\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/workspaces/rnacloud_genome_reference/misc/../rnacloud_genome_reference/common/gnomad.py:104\u001b[39m, in \u001b[36mGenomicRegionQuerier.query_multiple_regions\u001b[39m\u001b[34m(self, regions)\u001b[39m\n\u001b[32m    102\u001b[39m \u001b[38;5;66;03m# Use sync execution with session context\u001b[39;00m\n\u001b[32m    103\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m.client \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[32m--> \u001b[39m\u001b[32m104\u001b[39m     result = \u001b[43msession\u001b[49m\u001b[43m.\u001b[49m\u001b[43mexecute\u001b[49m\u001b[43m(\u001b[49m\u001b[43mquery_doc\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    106\u001b[39m logger.info(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mSuccessfully executed query for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mlen\u001b[39m(regions)\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m regions\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    108\u001b[39m \u001b[38;5;66;03m# Extract and organize results\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.local/lib/python3.12/site-packages/gql/client.py:1027\u001b[39m, in \u001b[36mSyncClientSession.execute\u001b[39m\u001b[34m(self, document, variable_values, operation_name, serialize_variables, parse_result, get_execution_result, **kwargs)\u001b[39m\n\u001b[32m   1025\u001b[39m \u001b[38;5;66;03m# Raise an error if an error is returned in the ExecutionResult object\u001b[39;00m\n\u001b[32m   1026\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m result.errors:\n\u001b[32m-> \u001b[39m\u001b[32m1027\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m TransportQueryError(\n\u001b[32m   1028\u001b[39m         str_first_element(result.errors),\n\u001b[32m   1029\u001b[39m         errors=result.errors,\n\u001b[32m   1030\u001b[39m         data=result.data,\n\u001b[32m   1031\u001b[39m         extensions=result.extensions,\n\u001b[32m   1032\u001b[39m     )\n\u001b[32m   1034\u001b[39m \u001b[38;5;28;01massert\u001b[39;00m (\n\u001b[32m   1035\u001b[39m     result.data \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m   1036\u001b[39m ), \u001b[33m\"\u001b[39m\u001b[33mTransport returned an ExecutionResult without data or errors\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m   1038\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m get_execution_result:\n",
      "\u001b[31mTransportQueryError\u001b[39m: {'message': 'Query is too expensive (44). Maximum allowed cost is 25.'}"
     ]
    }
   ],
   "source": [
    "with open(\"../temp/clinically_significant_protein_coding_genes.tsv\", 'r') as f:\n",
    "    reader = csv.DictReader(f, delimiter='\\t')\n",
    "\n",
    "    for row in reader:\n",
    "        if row['entrez_gene_id'] == '9636':\n",
    "            a = gtf_handler.obtain_sj_positions(\n",
    "                chrom=row['chrom_refseq'],\n",
    "                start=int(row['start']),\n",
    "                end=int(row['end']),\n",
    "                entrez_gene_id=int(row['entrez_gene_id'])\n",
    "            )\n",
    "\n",
    "            genomic_regions = [convert_sj_positions_to_genomic_region(sj) for sj in a]\n",
    "\n",
    "            print(genomic_regions)\n",
    "\n",
    "            gnomad_frequency = query_genomic_regions(genomic_regions)\n",
    "\n",
    "            break\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1552e301",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
