
# Gene Primary VS FIX contig comparison schema

## Output schema
* File is TSV formatted
* Has a header

| Field                        | Description                                                                                                                                   |
|------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|
| chr_refseq                   | Primary contig (RefSeq accession) for gene                                                                                                    |
| chr_ucsc                     | Primary contig (UCSC accession) for gene                                                                                                      |
| start                        | Primary contig start position for gene                                                                                                        |
| end                          | Primary contig end position for gene                                                                                                          |
| strand                       | Primary contig strand for gene                                                                                                                |
| gene_name                    | Gene name                                                                                                                                     |
| entrez_gene_id               | NCBI Gene ID / Entrez ID                                                                                                                      |
| issue_id                     | GRC Issue ID(s)                                                                                                                               |
| type                         | High-level classification of the problem affecting the gene (e.g. "Path Problem", "Gap")                                                      |
| summary                      | Short summary of the reference issue(s) for this gene, semi-colon separated if multiple                                                       |
| description                  | Extended description of reference issue(s) for this gene, semi-colon separated if multiple                                                    |
| alt_chr_refseq               | Fix contig (RefSeq accession)                                                                                                                 |
| alt_chr_ucsc                 | Fix contig (UCSC accession)                                                                                                                   |
| alt_scaf_start               | Fix contig start position                                                                                                                     |
| alt_scaf_stop                | Fix contig end position                                                                                                                       |
| primary_contig_transcript    | Transcript selected on the primary contig **<sup>1</sup>**                                                                                    |
| primary_contig_n_exons       | Number of exons in the primary contig transcript                                                                                              |
| primary_contig_n_introns     | Number of introns in the primary contig transcript                                                                                            |
| fix_contig_transcript        | Transcript selected on the fix contig; if none matches the primary transcript, blank                                                          |
| fix_contig_n_exons           | Number of exons in the fix contig transcript                                                                                                  |
| fix_contig_n_introns         | Number of introns in the fix contig transcript                                                                                                |
| n_exons_equal                | `True` if exon counts match between primary and fix; otherwise `False`                                                                        |
| n_introns_equal              | `True` if intron counts match between primary and fix; otherwise `False`                                                                      |
| sequences_unequal_n_exons    | Number of exons with sequence differences (see note **<sup>2</sup>**)                                                                         |
| sequences_unequal_n_introns  | Number of introns with sequence differences (see note **<sup>2</sup>**)                                                                       |
| splice_sites_unequal_n       | Number of splice sites (exon-intron boundaries) that differ between primary and fix                                                           |
| primary_exon_lengths         | Semi-colon separated list of exon lengths in the primary transcript                                                                           |
| primary_intron_lengths       | Semi-colon separated list of intron lengths in the primary transcript                                                                         |
| fix_exon_lengths             | Semi-colon separated list of exon lengths in the fix transcript                                                                               |
| fix_intron_lengths           | Semi-colon separated list of intron lengths in the fix transcript                                                                             |
| discordant_exon_numbering    | `True` if exon numbering between the two transcripts does not match (e.g. reversed or structural differences); otherwise `False`              |
| comparison_status            | See status descriptions in note **<sup>3</sup>**                                                                                              |
| clinically_relevant_gene     | `True` if the gene is on the clinically relevant list, otherwise `False`                                                                      |

**<sup>1</sup>** Primary contig transcript
- Preference is given to MANE
- Otherwise a non MANE transcript is selected (if only one non MANE transcript is present for the gene)

**<sup>2</sup>** Fix contig - No. of exons/introns
- If number of exons/introns differ between Primary/Fix, this is set to -1

**<sup>3</sup>** Comparison between gene on primary and FIX contig
- **Not comparable** - When either the fix contig transcript or primary contig transcript is missing (None)
- **Identical** - When all of the following conditions are met:
    - Number of exons are equal between transcripts
    - Number of introns are equal between transcripts
    - No exon sequences differ (sequences_unequal_n_exons = 0)
    - No intron sequences differ (sequences_unequal_n_introns = 0)
- **Different - No. of exons or introns differ** - When the number of exons or number of introns don't match between the two transcripts
- **Different - Sequences differ** - When the exon and intron counts match, but at least one exon sequence or intron sequence differs between the transcripts
- **Different - Splice-site sequences differ** - When the GT or the AG are different between Primary and Fix contigs
- **Different - Exon numbering is discordant** - If exon numbering is discordant due to strand reversal or differences in no. of exons in Primary vs Fix contig
- **Different - Unknown reason** - Fallback status for any other cases that don't match the above conditions

## Examples

| **chr_refseq** | **chr_ucsc** | **start** | **end**   | **strand** | **gene_name** | **entrez_gene_id** | **issue_id**          | **type**                    | **summary**                                                                                                                             | **description**                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | **alt_chr_refseq** | **alt_chr_ucsc**     | **alt_scaf_start** | **alt_scaf_stop** | **primary_contig_transcript** | **primary_contig_n_exons** | **primary_contig_n_introns** | **fix_contig_transcript** | **fix_contig_n_exons** | **fix_contig_n_introns** | **n_exons_equal** | **n_introns_equal** | **sequences_unequal_n_exons** | **sequences_unequal_n_introns** | **splice_sites_unequal_n** | **primary_exon_lengths**                                                                                  | **primary_intron_lengths**                                                                                                   | **fix_exon_lengths**                                                                                                                     | **fix_intron_lengths**                                                                                                                                           | **discordant_exon_numbering** | **comparison_status**                      | **clinically_relevant_gene** |
|----------------|--------------|-----------|-----------|------------|---------------|--------------------|-----------------------|-----------------------------|-----------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------|----------------------|--------------------|-------------------|-------------------------------|----------------------------|------------------------------|---------------------------|------------------------|--------------------------|-------------------|---------------------|-------------------------------|---------------------------------|----------------------------|-----------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------|--------------------------------------------|------------------------------|
| NC_000015.10   | chr15        | 21846329  | 21877735  | -          | POTEB         | 100996331          | HG-2365               | Path Problem                | A CH17 BAC pathway is being sequenced for the Prader-Willi region.                                                                      | A CH17 BAC pathway is currently being sequenced in the Prader-Willi region of Chr. 15.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             | NW_021160017.1     | chr15_ML143371v1_fix | 1                  | 5500449           | NM_001277304.2                | 11                         | 10                           | NM_001277304.2            | 11                     | 10                       | TRUE              | TRUE                | 11                            | 10                              | 20                         | 232;124;167;45;71;71;138;107;174;115;899                                                                  | 1638;8469;1045;1403;5493;2510;1337;2828;159;4382                                                                             | 884;115;174;107;138;71;71;45;167;124;232                                                                                                 | 4368;159;2823;1337;2508;5493;1403;1045;8445;1639                                                                                                                 | TRUE                          | Different - Exon numbering is discordant   | FALSE                        |
| NC_000001.11   | chr1         | 16562423  | 16613564  | -          | NBPF1         | 55672              | HG-1343;HG-173;HG-459 | Gap;Missing sequence;Gap    | Chr. 1 CHORI-17 Pathway: MST1P2 (Gene ID: 11209);GeneID: 55672 (NBPF1), poor quality (includes component misassembly);Chr.1 Gap Problem | Clones from the CH17 library have been selected for sequencing that align near the AL358794.19 - BX284668.4 region of Chr. 1 (GRCh38 coordinates NC_000001.11:16,310,000-17,150,000). These will be used to close a sequence gap in the region and/or represent an alternate haplotype.;There is poor alignment between Reference component AL355149.13 and NBPF1 (GeneID:55672/NM_017940) possibly due to repetitive nature of the region or the sequence representing NBPF1 is still missing from the Reference Assembly.;There is a gap in the Reference Assembly between components AL021920.1 and BX284668.4. | NW_025791756.1     | chr1_MU273333v1_fix  | 1                  | 1572686           | NM_001405678.2                | 28                         | 27                           | NM_001405678.2            | 37                     | 36                       | FALSE             | FALSE               | -1                            | -1                              | -1                         | 1764;112;175;181;52;164;52;206;73;215;103;210;212;73;215;103;210;212;73;215;103;210;155;127;70;83;272;227 | 620;714;1363;627;1042;3905;1293;464;1037;835;1773;1342;463;1051;835;1792;1351;463;1051;835;1845;102;1127;1024;269;13494;4558 | 227;272;83;70;127;155;210;103;215;73;212;210;103;215;73;212;210;103;215;73;206;52;164;52;173;52;173;52;173;112;173;52;80;52;173;112;1764 | 4554;13515;269;1022;1127;102;1845;835;1051;463;1346;1795;835;1050;463;1349;1773;835;1037;464;1291;3906;1042;639;719;602;709;622;714;606;719;602;802;616;714;5327 |                               | Different - No. of exons or introns differ | FALSE                        |
| NC_000001.11   | chr1         | 16246840  | 16352480  | -          | FBXO42        | 54455              | HG-1343;HG-173;HG-459 | Gap;Missing sequence;Gap    | Chr. 1 CHORI-17 Pathway: MST1P2 (Gene ID: 11209);GeneID: 55672 (NBPF1), poor quality (includes component misassembly);Chr.1 Gap Problem | Clones from the CH17 library have been selected for sequencing that align near the AL358794.19 - BX284668.4 region of Chr. 1 (GRCh38 coordinates NC_000001.11:16,310,000-17,150,000). These will be used to close a sequence gap in the region and/or represent an alternate haplotype.;There is poor alignment between Reference component AL355149.13 and NBPF1 (GeneID:55672/NM_017940) possibly due to repetitive nature of the region or the sequence representing NBPF1 is still missing from the Reference Assembly.;There is a gap in the Reference Assembly between components AL021920.1 and BX284668.4. | NW_025791756.1     | chr1_MU273333v1_fix  | 1                  | 1572686           | NM_018994.3                   | 10                         | 9                            | NM_018994.3               | 10                     | 9                        | TRUE              | TRUE                | 0                             | 1                               | 0                          | 4946;117;57;97;111;154;135;117;267;226                                                                    | 502;691;482;1979;784;38023;10885;9249;36819                                                                                  | 4946;117;57;97;111;154;135;117;267;226                                                                                                   | 502;691;482;1979;784;38023;10885;9249;36825                                                                                                                      | FALSE                         | Different - Sequences differ               | FALSE                        |
| NC_000013.11   | chr13        | 113926514 | 113974076 | -          | C13orf46      | 100507747          | HG-2288;HG-2289       | Clone Problem;Clone Problem | Restoring component dropped from GRCh38;Alignment problem caused by AUXG01000135.1                                                      | Component FP565325.3 inadvertently dropped from GRCh38 restored to TPF;Overlap between BX537329.2 and AL161774.49 has been split by the addition of WGS contig AUXG01000135.1 causing and artificial duplication in the reference assembly.                                                                                                                                                                                                                                                                                                                                                                        | NW_011332698.1     | chr13_KN538371v1_fix | 1                  | 206320            | NM_001365455.2                | 7                          | 6                            | NM_001365455.2            | 7                      | 6                        | TRUE              | TRUE                | 0                             | 3                               | 1                          | 3135;68;48;48;166;52;269                                                                                  | 8087;2346;1078;80;1410;3585                                                                                                  | 3135;68;48;48;166;52;269                                                                                                                 | 8087;2346;1079;80;1410;3585                                                                                                                                      | FALSE                         | Different - Splice-site sequences differ   | FALSE                        |
| NC_000001.11   | chr1         | 12824605  | 12831410  | -          | PRAMEF11      | 440560             | HG-1342;HG-2282       | Path Problem;Gap            | Chr. 1 CHORI-17 Pathway: PRAMEF Genes;There is a gap in the Reference between components AC245034.2 and AC245056.3 in GRCh38.           | Clones have been selected for sequencing that align near the BX682532.4 - AL354712.18 region of Chr. 1. These may be used to improve the reference or to represent an alternate haplotype.;There is a gap in the Reference between components AC245034.2 and AC245056.3 in GRCh38.  This gap ticket is a continuation of the same gap outlined in HG-457.                                                                                                                                                                                                                                                          | NW_012132914.1     | chr1_KQ031383v1_fix  | 1                  | 467143            | NM_001146344.3                | 4                          | 3                            | NM_001146344.3            | 4                      | 3                        | TRUE              | TRUE                | 0                             | 0                               | 0                          | 899;582;309;55                                                                                            | 1745;666;2550                                                                                                                | 899;582;309;55                                                                                                                           | 1745;666;2550                                                                                                                                                    | FALSE                         | Identical                                  | FALSE                        |
| NC_000001.11   | chr1         | 16538986  | 16539663  | +          | LOC124903856  | 124903856          | HG-1343;HG-173;HG-459 | Gap;Missing sequence;Gap    | Chr. 1 CHORI-17 Pathway: MST1P2 (Gene ID: 11209);GeneID: 55672 (NBPF1), poor quality (includes component misassembly);Chr.1 Gap Problem | Clones from the CH17 library have been selected for sequencing that align near the AL358794.19 - BX284668.4 region of Chr. 1 (GRCh38 coordinates NC_000001.11:16,310,000-17,150,000). These will be used to close a sequence gap in the region and/or represent an alternate haplotype.;There is poor alignment between Reference component AL355149.13 and NBPF1 (GeneID:55672/NM_017940) possibly due to repetitive nature of the region or the sequence representing NBPF1 is still missing from the Reference Assembly.;There is a gap in the Reference Assembly between components AL021920.1 and BX284668.4. | NW_025791756.1     | chr1_MU273333v1_fix  | 1                  | 1572686           | XM_047436898.1                | 1                          | 0                            |                           | 0                      | 0                        | FALSE             | TRUE                | -1                            | 0                               | 0                          | 678                                                                                                       |                                                                                                                              |                                                                                                                                          |                                                                                                                                                                  |                               | Not comparable                             | FALSE                        |
