
# GRC fixes assessment

## Output schema
* File is TSV formatted
* Has a header

| Field                                    | Description                                                                                                                                                     |
| ---------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| chr_refseq                               | Primary contig (RefSeq accession) for gene                                                                                                                      |
| chr_ucsc                                 | Primary contig (UCSC accession) for gene                                                                                                                        |
| start                                    | Primary contig start position for gene                                                                                                                          |
| end                                      | Primary contig end position for gene                                                                                                                            |
| strand                                   | Primary contig strand for gene                                                                                                                                  |
| gene_name                                | Gene name                                                                                                                                                       |
| entrez_gene_id                           | NCBI Gene ID / Entrez ID                                                                                                                                        |
| issue_id                                 | GRC Issue ID(s)                                                                                                                                                 |
| type                                     | High-level classification of the problem affecting the gene (e.g. "Path Problem", "Gap")                                                                        |
| summary                                  | Short summary of the reference issue(s) for this gene, semi-colon separated if multiple                                                                         |
| description                              | Extended description of reference issue(s) for this gene, semi-colon separated if multiple                                                                      |
| alt_chr_refseq                           | Fix contig (RefSeq accession)                                                                                                                                   |
| alt_chr_ucsc                             | Fix contig (UCSC accession)                                                                                                                                     |
| alt_scaf_start                           | Fix contig start position                                                                                                                                       |
| alt_scaf_stop                            | Fix contig end position                                                                                                                                         |
| primary_contig_transcript                | Transcript selected on the primary contig **<sup>1</sup>**                                                                                                      |
| primary_contig_transcript_is_mane_select | True if primary contig transcript is MANE Select, otherwise False                                                                                               |
| primary_contig_transcript_partial        | True if Primary contig is partial (This can happen when Refseq sequence does not perfect align with the assembly e.g. SHANK3 / GRCh38.p14 / chr22)              |
| primary_contig_n_exons                   | Number of exons in the primary contig transcript                                                                                                                |
| primary_contig_n_introns                 | Number of introns in the primary contig transcript                                                                                                              |
| fix_contig_transcript                    | Transcript selected on the fix contig; if none matches the primary transcript, blank                                                                            |
| fix_contig_transcript_is_mane_select     | True if Fix contig transcript is MANE Select, otherwise False                                                                                                   |
| fix_contig_transcript_partial            | True if Fix contig is partial (This can happen when the fix annotation does not span the entire gene e.g. TBC1D16 / HG-2118 / GRCh38.p14  chr17_MU273379v1_fix) |
| fix_contig_n_exons                       | Number of exons in the fix contig transcript                                                                                                                    |
| fix_contig_n_introns                     | Number of introns in the fix contig transcript                                                                                                                  |
| n_exons_equal                            | `True` if exon counts match between primary and fix; otherwise `False`                                                                                          |
| n_introns_equal                          | `True` if intron counts match between primary and fix; otherwise `False`                                                                                        |
| sequences_unequal_n_exons                | Number of exons with sequence differences (see note **<sup>2</sup>**)                                                                                           |
| sequences_unequal_n_introns              | Number of introns with sequence differences (see note **<sup>2</sup>**)                                                                                         |
| splice_sites_unequal_n                   | Number of splice sites (exon-intron boundaries) that differ between primary and fix                                                                             |
| primary_exon_lengths                     | Semi-colon separated list of exon lengths in the primary transcript                                                                                             |
| primary_intron_lengths                   | Semi-colon separated list of intron lengths in the primary transcript                                                                                           |
| fix_exon_lengths                         | Semi-colon separated list of exon lengths in the fix transcript                                                                                                 |
| fix_intron_lengths                       | Semi-colon separated list of intron lengths in the fix transcript                                                                                               |
| discordant_exon_numbering                | `True` if exon numbering between the two transcripts does not match (e.g. reversed or structural differences); otherwise `False`                                |
| comparison_status                        | See status descriptions in note **<sup>3</sup>**                                                                                                                |
| clinically_relevant_gene                 | `True` if the gene is on the clinically relevant list, otherwise `False`                                                                                        |

**<sup>1</sup>** Primary contig transcript
- Preference is given to MANE
- Otherwise a non MANE transcript is selected (if only one non MANE transcript is present for the gene)

**<sup>2</sup>** Fix contig - No. of exons/introns
- If number of exons/introns differ between Primary/Fix, this is set to -1

**<sup>3</sup>** Comparison between gene on primary and FIX contig
- **Not comparable** - When either the fix contig transcript or primary contig transcript is missing (None)
- **Identical** - When all of the following conditions are met:
    - Number of exons are equal between transcripts
    - Number of introns are equal between transcripts
    - No exon sequences differ (sequences_unequal_n_exons = 0)
    - No intron sequences differ (sequences_unequal_n_introns = 0)
- **Different - No. of exons or introns differ** - When the number of exons or number of introns don't match between the two transcripts
- **Different - Sequences differ** - When the exon and intron counts match, but at least one exon sequence or intron sequence differs between the transcripts
- **Different - Splice-site sequences differ** - When the GT or the AG are different between Primary and Fix contigs
- **Different - Exon numbering is discordant** - If exon numbering is discordant due to strand reversal or differences in no. of exons in Primary vs Fix contig
- **Different - Unknown reason** - Fallback status for any other cases that don't match the above conditions

## Examples

| chr_refseq   | chr_ucsc             | start     | end       | strand | gene_name | entrez_gene_id | issue_id        | type                        | summary                                                                                                                                      | description                                                                                                                                                                                                                                                                                                                          | alt_chr_refseq | alt_chr_ucsc         | alt_scaf_start | alt_scaf_stop | primary_contig_transcript | primary_contig_transcript_is_mane_select | primary_contig_transcript_partial | primary_contig_n_exons | primary_contig_n_introns | fix_contig_transcript | fix_contig_transcript_is_mane_select | fix_contig_transcript_partial | fix_contig_n_exons | fix_contig_n_introns | n_exons_equal | n_introns_equal | sequences_unequal_n_exons | sequences_unequal_n_introns | splice_sites_unequal_n | primary_exon_lengths                                                                                                | primary_intron_lengths                                                                                                     | fix_exon_lengths                                                                                                    | fix_intron_lengths                                                                                                         | discordant_exon_numbering | comparison_status                                          | clinically_relevant_gene |
|--------------|----------------------|-----------|-----------|--------|-----------|----------------|-----------------|-----------------------------|----------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------|----------------------|----------------|---------------|---------------------------|------------------------------------------|-----------------------------------|------------------------|--------------------------|-----------------------|--------------------------------------|-------------------------------|--------------------|----------------------|---------------|-----------------|---------------------------|-----------------------------|------------------------|---------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------|---------------------------|------------------------------------------------------------|--------------------------|
| NC_000009.12 | chr9                 | 133250401 | 133275201 | -      | ABO       | 28             | HG-2030         | Path Problem                | Non-functional ABO allele represented in GRCh38                                                                                              | GRCh38 was supposed to represent a functional ABO*A1.02 allele contained solely within AL772161.10, as represented in the NW_003315925.1 fix patch and discussed in JIRA HG-79. However, GRCh38 (NC_000009.12) has a different tiling path compared to the fix patch, and a non-functional O-type ABO allele is instead represented. | NW_009646201.1 | chr9_KN196479v1_fix  | 1              | 330164        | NM_020469.3               | FALSE                                    | FALSE                             | 8                      | 7                        | NM_020469.3           | FALSE                                | FALSE                         | 7                  | 6                    | FALSE         | FALSE           | -1                        | -1                          | -1                     | 5956;113;19;36;48;57;70;40                                                                                          | 1052;2;554;1686;1451;724;12993                                                                                             | 6092;135;36;48;57;70;40                                                                                             | 1052;554;1686;1451;724;12993                                                                                               |                           | Different - No. of exons or introns differ                 | FALSE                    |
| NC_000022.11 | chr22                | 50738204  | 50745339  | +      | ACR       | 49             | HG-1311;HG-2539 | Clone Problem;Clone Problem | Deletion in genome in coding exon of SHANK3;GeneID: 85358 (SHANK3), poor quality (includes component misassembly)                            | Around 42 bp of SHANK3 cDNA AB569469.1 do not match the genome. Missing GC-rich sequence is from a CpG island that contains exon 12 of CCDS46737.;In progress                                                                                                                                                                        | NW_015148969.2 | chr22_KQ759762v2_fix | 1              | 101040        | NM_001097.3               | TRUE                                     | FALSE                             | 5                      | 4                        | NM_001097.3           | TRUE                                 | FALSE                         | 5                  | 4                    | TRUE          | TRUE            | 0                         | 1                           | 0                      | 109;204;284;146;687                                                                                                 | 958;219;4083;446                                                                                                           | 109;204;284;146;687                                                                                                 | 958;219;4077;446                                                                                                           | FALSE                     | Different - Sequences differ                               | FALSE                    |
| NC_000017.11 | chr17                | 81509971  | 81512799  | -      | ACTG1     | 71             | HG-1369         | Clone Problem               | AC213194.1 is a Phase 2 Clone in the Reference                                                                                               | Sequence from phase 2 clone AC213194.1 should be replaced with finished sequence.                                                                                                                                                                                                                                                    | NW_025791805.1 | chr17_MU273382v1_fix | 1              | 187626        | NM_001614.5               | TRUE                                     | FALSE                             | 6                      | 5                        | NM_001614.5           | TRUE                                 | FALSE                         | 6                  | 5                    | TRUE          | TRUE            | 0                         | 0                           | 0                      | 863;182;439;240;129;66                                                                                              | 93;79;276;89;373                                                                                                           | 863;182;439;240;129;66                                                                                              | 93;79;276;89;373                                                                                                           | FALSE                     | Identical                                                  | TRUE                     |
| NC_000019.10 | chr19                | 38647649  | 38731589  | +      | ACTN4     | 81             | HG-26           | Clone Problem               | Hsa19- CAPN12/ECH1 problem region                                                                                                            | ESP analysis suggests the reference chromosome contains an inversion between AC022144.8 and AC008982.5                                                                                                                                                                                                                               | NW_014040929.1 | chr19_KQ458386v1_fix | 1              | 405389        | NM_004924.6               | TRUE                                     | FALSE                             | 21                     | 20                       | NM_004924.6           | TRUE                                 | FALSE                         | 21                 | 20                   | TRUE          | TRUE            | 0                         | 0                           | 0                      | 259;115;120;87;88;79;82;86;93;231;148;151;109;141;183;135;180;147;81;159;2316                                       | 52692;287;3812;1023;1985;1199;780;4126;2524;610;3463;1925;214;79;91;1158;1053;842;969;119                                  | 259;115;120;87;88;79;82;86;93;231;148;151;109;141;183;135;180;147;81;159;2316                                       | 52692;287;3812;1023;1985;1199;780;4126;2524;610;3463;1925;214;79;91;1158;1053;842;969;119                                  | FALSE                     | Identical                                                  | TRUE                     |
| NT_187660.1  | chr15_KI270905v1_alt | 1196581   | 1393362   | +      | APBA2     | 321            | HG-2139         | Missing sequence            | Sequence needs added to Chr. 15 GRCh38 Alt_Loci contig NT_187660.1.                                                                          | Sequence needs added to Chr. 15 GRCh38 Alt_Loci contig NT_187660.1.                                                                                                                                                                                                                                                                  | NW_011332701.1 | chr15_KN538374v1_fix | 1              | 4998962       | NM_001379685.1            | FALSE                                    | TRUE                              | 14                     | 13                       | NM_001379685.1        | FALSE                                | FALSE                         | 16                 | 15                   | FALSE         | FALSE           | -1                        | -1                          | -1                     | 124;54;991;81;37;146;36;87;186;180;213;120;141;1258                                                                 | 73998;58046;20168;1053;16983;1057;4172;3021;3603;1048;1450;5486;3043                                                       | 305;90;110;54;991;81;37;146;36;87;186;180;213;120;141;1258                                                          | 9240;26520;73998;58046;20168;1053;16983;1057;4172;3021;3603;1048;1450;5486;3043                                            |                           | Not comparable - Partial transcript annotation in GTF file | FALSE                    |
| NC_000014.9  | chr14                | 20455226  | 20457767  | +      | APEX1     | 328            | HG-2526;HG-2573 | Variation;Variation         | GeneID: 390432 (OR4Q2), poor quality (includes component misassembly);GeneID: 390441 (OR11H7), poor quality (includes component misassembly) | In progress;GeneID: 390441 (OR11H7) aligns poorly to the Reference.                                                                                                                                                                                                                                                                  | NW_025791796.1 | chr14_MU273373v1_fix | 1              | 722645        | NM_001641.4               | TRUE                                     | FALSE                             | 5                      | 4                        | NM_001641.4           | TRUE                                 | FALSE                         | 5                  | 4                    | TRUE          | TRUE            | 0                         | 0                           | 0                      | 169;126;188;193;777                                                                                                 | 183;210;566;130                                                                                                            | 169;126;188;193;777                                                                                                 | 183;210;566;130                                                                                                            | FALSE                     | Identical                                                  | FALSE                    |
| NC_000002.12 | chr2                 | 21001429  | 21044073  | -      | APOB      | 338            | HG-2231;HG-2496 | Clone Problem;Variation     | Replace IUPAC ambiguity bases in GRCh38;GRCh38 represents rare allele in APOB (rs584542)                                                     | Ambiguous IUPAC base (Y) in NC_000002.12: 20953875 in component AEKP01019831.1;The GRCh38 NC_000002.12: 21009931 (dbSNP: rs584542) is observed with global MAF=0.0138 in the 1000G Phase 3 May 2013 call set.                                                                                                                        | NW_025791767.1 | chr2_MU273344v1_fix  | 1              | 244725        | NM_000384.3               | TRUE                                     | FALSE                             | 29                     | 28                       | NM_000384.3           | TRUE                                 | FALSE                         | 29                 | 28                   | TRUE          | TRUE            | 1                         | 0                           | 0                      | 1906;184;115;7572;374;146;188;176;211;122;183;212;168;192;177;238;212;147;118;228;220;86;125;156;154;146;116;39;210 | 934;108;404;508;914;479;109;893;2342;609;2925;482;1240;1663;863;261;1100;112;2338;717;1297;682;1391;702;2826;1277;1036;312 | 1906;184;115;7572;374;146;188;176;211;122;183;212;168;192;177;238;212;147;118;228;220;86;125;156;154;146;116;39;210 | 934;108;404;508;914;479;109;893;2342;609;2925;482;1240;1663;863;261;1100;112;2338;717;1297;682;1391;702;2826;1277;1036;312 | FALSE                     | Different - Sequences differ                               | TRUE                     |
